### PR TITLE
[Doc] [refactor] Improve Taichi kernels and functions definition

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: docs/syntax.rst [hunk 1]
```diff
@@ -1,34 +1,111 @@
...
+       ...
+
+   my_kern()
+
+
+Kernels are to be called from **Python-scope**.
+
+.. note::
+
+    For people from CUDA, Taichi kernels = ``__global__``.
```
Comment (TroyZhai) [inline docs/syntax.rst:45]:
```suggestion
    For people from CUDA, Taichi kernels = ``__global__`` functions.
```

File: docs/syntax.rst [hunk 2]
```diff
@@ -1,34 +1,111 @@
...
+        def good_kernel(vx: ti.f32, vy: ti.f32):
+            v = ti.Vector([vx, vy])
+            ...
+
+
+Return value
+************
+
+A kernel may or may not have a **scalar** return value.
+The type of return value must be hinted:
```
Comment (TroyZhai) [inline docs/syntax.rst:84]:
```suggestion
If it does, the type of return value must be hinted:
```

File: docs/syntax.rst [hunk 3]
```diff
@@ -1,34 +1,111 @@
...
 
-    res = add_xy(2.3, 1.1)
-    print(res)  # 3, since return type is ti.i32
+    print(my_kern())  # 233, since return type is ti.i32
 
 
 .. note::
 
-    For now, we only support one scalar as return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. For example:
+    For now, kernels only support one scalar as return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. For example:
```
Comment (TroyZhai) [inline docs/syntax.rst:108]:
```suggestion
    For now, a kernel can only have one scalar return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. For example:
```

File: docs/syntax.rst [hunk 4]
```diff
@@ -43,74 +120,111 @@ The return value will be automatically cast into the hinted type. e.g.,
...
+        ...
+        my_func()  # call functions from Taichi-scope
+        ...
 
+    my_kern()      # call kernels from Python-scope
 
-.. warning::
 
-    Functions with multiple ``return`` statements are not supported for now. Use a **local** variable to store the results, so that you end up with only one ``return`` statement:
+Functions are to be called from **Taichi-scope**.
```
Comment (TroyZhai) [inline docs/syntax.rst:155]: I think we'd better differentiate the term functions and taichi functions.

File: docs/syntax.rst [hunk 5]
```diff
@@ -43,74 +120,111 @@ The return value will be automatically cast into the hinted type. e.g.,
...
+.. note::
 
-      # Bad function - two return statements
-      @ti.func
-      def safe_sqrt(x):
-        if x >= 0:
-          return ti.sqrt(x)
-        else:
-          return 0.0
+    For people from CUDA, Taichi functions = ``__device__``.
```
Comment (TroyZhai) [inline docs/syntax.rst:159]:
```suggestion
    For people from CUDA, Taichi functions = ``__device__`` functions.
```

File: docs/syntax.rst [hunk 6]
```diff
@@ -43,74 +120,111 @@ The return value will be automatically cast into the hinted type. e.g.,
...
+    Functions can be nested.
 
-    Currently, all functions are force-inlined. Therefore, no recursion is allowed.
 
+Arguments and return values
+***************************
 
-.. note::
+Functions can have multiple arguments and return values.
+Unlike kernel, function arguments don't need to be type-hinted:
```
Comment (TroyZhai) [inline docs/syntax.rst:170]:
```suggestion
Unlike kernels, arguments in functions don't need to be type-hinted:
```

File: docs/type.rst [hunk 7]
```diff
@@ -1,8 +1,33 @@
 Type system
 ===========
 
+Taichi supports many kind of data types, the type name is recognized as
```
Comment (TroyZhai) [inline docs/type.rst:4]: 'many' is not compatible with 'kind' here.
```suggestion
Taichi supports many data types. The type name is recognized as
```

File: docs/type.rst [hunk 8]
```diff
@@ -1,8 +1,33 @@
...
+- ``f`` for floating point numbers, e.g. 2.33, 1e-4
+
+The *digital number* can be one of:
+
+- ``8``
+- ``16``
+- ``32``
+- ``64``
+
+It represents how much **bits** is used in storing the data.
```
Comment (TroyZhai) [inline docs/type.rst:20]:
```suggestion
It represents how many **bits** are used in storing the data.
```

File: docs/type.rst [hunk 9]
```diff
@@ -1,8 +1,33 @@
...
+The larger the bit number, the higher precision is.
```
Comment (TroyZhai) [inline docs/type.rst:21]:
```suggestion
The larger the bit number, the higher the precision is.
```

File: docs/type.rst [hunk 10]
```diff
@@ -46,13 +71,21 @@ Currently, supported basic types in Taichi are
...
 
-- ``i32 + f32 = f32``
-- ``f32 + f64 = f64``
-- ``i32 + i64 = i64``
+
+
+Type promotion
+--------------
+
+Binary operations on different types will give you a promoted type, following the C programming language, e.g.:
```
Comment (TroyZhai) [inline docs/type.rst:83]:
```suggestion
Binary operations on different types will give you a promoted type, following the C programming language convention, e.g.:
```

File: docs/type.rst [hunk 11]
```diff
@@ -46,13 +71,21 @@ Currently, supported basic types in Taichi are
...
+
+- ``i32 + f32 = f32`` (integer + float = float)
+- ``i32 + i64 = i64`` (less-bits + more-bits = lamore-bits)
+
+Basically it will try to choose the least precise type to contain the result value.
```
Comment (TroyZhai) [inline docs/type.rst:88]: Umm... I'm not following here. Do you mean 'most'?

File: docs/type.rst [hunk 12]
```diff
@@ -76,24 +110,64 @@ Default integer and float-point precisions (``default_ip`` and ``default_fp``) c
...
+.. code-block:: python
+
+    a = 1.7
+    b = int(a)    # 1
+    c = float(a)  # 1.0
+
+Casting vector / matrix elements
+********************************
+
+Type cast applied to vectors / matrices are element-wise:
```
Comment (TroyZhai) [inline docs/type.rst:158]: I think the slash here doesn't need the spaces before and after. But please don't quote me here until we get opinions from a native speaker.
```suggestion
Type casts applied to vectors/matrices are element-wise:
```

File: docs/syntax.rst [hunk 13]
```diff
@@ -126,11 +240,41 @@ Use ``@ti.func`` to decorate your Taichi functions. These functions are callable
...
+          ret = ti.sqrt(x)
+        else:
+          ret = 0.0
+        return ret
+
+
+.. warning::
+
+    Currently, all functions are force-inlined. Therefore, no recursion is allowed.
+
```
Comment (archibate) [inline docs/syntax.rst:271]:
```suggestion
```

File: docs/syntax.rst (hunk 4 above)
Comment (archibate) [reply docs/syntax.rst:155]:
```suggestion
Taichi functions can only be called from **Taichi-scope**.
```

File: docs/syntax.rst [hunk 14]
```diff
@@ -1,34 +1,111 @@
...
+.. code-block:: python
+
+   @ti.kernel
+   def my_kern():
+       ...
+
+   my_kern()
+
+
+Kernels are to be called from **Python-scope**.
```
Comment (archibate) [inline docs/syntax.rst:41]:
```suggestion
Kernels should be called from **Python-scope**.
```

File: docs/syntax.rst [hunk 15]
```diff
@@ -1,34 +1,111 @@
 Syntax
 ======
 
+Taichi-scope vs Python-scope
+----------------------------
+
+Codes decorated by ``@ti.kernel`` or ``@ti.func`` is the **Taichi-scope**.
```
[the remark this answers is not in the document]
Comment (archibate) [reply docs/syntax.rst:7]:
```suggestion
Code decorated by ``@ti.kernel`` or ``@ti.func`` is in the **Taichi-scope**.
```

File: docs/syntax.rst [hunk 16]
```diff
@@ -1,34 +1,111 @@
...
+
+They are to be compiled and executed on CPU or GPU devices with high
+parallelization performance, on the cost of less flexibility.
+
+.. note::
+
+    For people from CUDA, Taichi-scope = **device** side.
+
+
+Codes outside ``@ti.kernel`` or ``@ti.func`` is the **Python-scope**.
```
[the remark this answers is not in the document]
Comment (archibate) [reply docs/syntax.rst:17]:
```suggestion
Code outside ``@ti.kernel`` or ``@ti.func`` is in the **Python-scope**.
```

File: docs/hello.rst [hunk 17]
```diff
@@ -101,33 +106,35 @@ Taichi programs run on either CPUs or GPUs. Initialize Taichi according to your
...
+  **Taichi-scopes v.s. Python-scopes**: everything decorated with ``@ti.kernel`` and ``@ti.func`` is in Taichi-scope, which will be compiled by the Taichi compiler.
+
+  Everything else is in Python-scopes. They are simply Python native code.
 
 .. warning::
 
   Taichi kernels must be called in the Python-scope. I.e., **nested kernels are not supported**.
-  Nested functions are allowed. **Recursive functions are not supported for now**.
+
+  But nested functions are allowed. **Recursive functions are not supported for now**.
```
Comment (archibate) [inline docs/hello.rst:137]:
```suggestion
  Nested functions are allowed. **Recursive functions are not supported for now**.
```

File: docs/syntax.rst [hunk 18]
```diff
@@ -43,74 +120,111 @@ The return value will be automatically cast into the hinted type. e.g.,
...
-        rst = 0.0
-        if x >= 0:
-          rst = ti.sqrt(x)
-        else:
-          rst = 0.0
-        return rst
+.. note::
 
-.. warning::
+    Functions can be nested.
```
Comment (archibate) [inline docs/syntax.rst:163]:
```suggestion
    Taichi functions can be nested.
    
.. warning::

    Currently, all functions are force-inlined. Therefore, no recursion is allowed.
```

File: docs/type.rst [hunk 19]
```diff
@@ -76,24 +110,64 @@ Default integer and float-point precisions (``default_ip`` and ``default_fp``) c
...
+
+Casting vector / matrix elements
+********************************
+
+Type casts applied to vectors/matrices are element-wise:
+
+.. code-block:: python
+
+    u = ti.Vector([2.3, 4.7])
+    v = int(u)  # ti.Vector([2, 4])
```
Comment (archibate) [inline docs/type.rst:163]:
```suggestion
    v = int(u)  # ti.Vector([2, 4])
    # equivalent to:
    v = ti.cast(u, ti.i32)  # ti.Vector([2, 4])
```

File: docs/hello.rst [hunk 20]
```diff
@@ -101,28 +102,29 @@ Taichi programs run on either CPUs or GPUs. Initialize Taichi according to your
...
 
 Functions and kernels
 ---------------------
 
 Computation resides in Taichi **kernels**. Kernel arguments must be type-hinted.
-The language used in Taichi kernels and functions looks exactly like Python, yet the Taichi frontend compiler converts it
+The language used in Taichi kernels and functions looks much like Python, yet the Taichi frontend compiler converts it
 into a language that is **compiled, statically-typed, lexically-scoped, parallel and differentiable**.
 
 Taichi **functions**, which can be called by Taichi kernels and other Taichi functions, should be defined with the keyword ``ti.func``.
```
Comment (archibate) [inline docs/hello.rst:121]:
```suggestion
Taichi **functions** are defined with the decorator ``@ti.func``.
They can be called by Taichi kernels and other Taichi functions.
```

File: docs/hello.rst [hunk 21]
```diff
@@ -101,28 +102,29 @@ Taichi programs run on either CPUs or GPUs. Initialize Taichi according to your
...
 
 In the code above, ``pixels = ti.var(dt=ti.f32, shape=(n * 2, n))`` allocates a 2D dense tensor named ``pixels`` of
-size ``(640, 320)`` and element data type ``ti.f32`` (i.e. ``float`` in C).
+size ``(640, 320)`` with data type ``ti.f32`` (32-bit floating-point number).
+
 
 Functions and kernels
 ---------------------
 
 Computation resides in Taichi **kernels**. Kernel arguments must be type-hinted.
```
Comment (archibate) [inline docs/hello.rst:117]:
```suggestion
Computation resides in Taichi **kernels**, which is defined with the decorator ``@ti.kernel``.
Kernel arguments must be type-hinted (if any).
```

File: docs/hello.rst [hunk 22]
```diff
@@ -109,32 +108,35 @@ Initialize Taichi according to your hardware platform as follows:
...
-size ``(640, 320)`` with data type ``ti.f32`` (32-bit floating-point number).
+size ``(640, 320)`` and element data type ``ti.f32`` (i.e. ``float`` in C).
 
 
 Functions and kernels
 ---------------------
 
-Computation resides in Taichi **kernels**. Kernel arguments must be type-hinted.
-The language used in Taichi kernels and functions looks much like Python, yet the Taichi frontend compiler converts it
+Computation resides in Taichi **kernels**, which is defined with the decorator ``@ti.kernel``.
```
Comment (TroyZhai) [inline docs/hello.rst:122]:
```suggestion
Computation resides in Taichi **kernels** which are defined with the decorator ``@ti.kernel``.
```

File: docs/hello.rst [hunk 23]
```diff
@@ -109,32 +108,35 @@ Initialize Taichi according to your hardware platform as follows:
...
+Kernel arguments must be type-hinted (if any).
+The language used in Taichi kernels and functions looks exactly like Python, yet the Taichi frontend compiler converts it
 into a language that is **compiled, statically-typed, lexically-scoped, parallel and differentiable**.
 
-Taichi **functions**, which can be called by Taichi kernels and other Taichi functions, should be defined with the keyword ``ti.func``.
+Taichi **functions** are defined with the decorator ``@ti.func``.
+They can be called by Taichi kernels and other Taichi functions.
```
Comment (TroyZhai) [inline docs/hello.rst:128]: Nit: you could join the above 2 sentences with 'and' and replace the 'and' there with 'or'.

File: docs/hello.rst [hunk 24]
```diff
@@ -113,25 +114,42 @@ size ``(640, 320)`` and element data type ``ti.f32`` (i.e. ``float`` in C).
...
 .. note::
 
-  **Taichi-scopes v.s. Python-scopes**: everything decorated with ``ti.kernel`` and ``ti.func`` is in Taichi-scope, which will be compiled by the Taichi compiler.
-  Everything else is in Python-scopes. They are simply Python code.
+  **Taichi-scopes v.s. Python-scopes**:
+
+  Everything decorated with ``@ti.kernel`` and ``@ti.func`` is in Taichi-scope
+  and hence will be compiled by the Taichi compiler.
+
+  Everything else is in Python-scopes. They are simply Python native code.
```
Comment (isdanni) [inline docs/hello.rst:137]: Shall we unify the format for `Python-scope` and `Taichi-scope`? If one is using plural, the other one should too.

File: docs/hello.rst [hunk 25]
```diff
@@ -113,25 +114,42 @@ size ``(640, 320)`` and element data type ``ti.f32`` (i.e. ``float`` in C).
...
+.. note::
 
-  Taichi functions can only be called in Taichi-scope.
+    For those who come from the world of CUDA, ``ti.func`` corresponds to ``__device__`` while ``ti.kernel`` corresponds to ``__global__``.
+
+.. warning::
 
-For those who come from the world of CUDA, ``ti.func`` corresponds to ``__device__`` while ``ti.kernel`` corresponds to ``__global__``.
+  **Nested kernels are not supported**.
+  **Nested functions are allowed**.
```
Comment (isdanni) [inline docs/hello.rst:151]:
```suggestion
  **Nested functions are supported**.
```
Maybe align with the other two lines?

File: docs/syntax.rst [hunk 26]
```diff
@@ -43,74 +120,118 @@ The return value will be automatically cast into the hinted type. e.g.,
...
+    @ti.kernel
+    def my_kernel():
+        ...
+        ret = my_add(2, 3.3)
+        print(ret)  # 5.3
+        ...
+
+
+Function arguments are passed by value, changes made inside function scope
+won't affect the outside value in the caller:
```
Comment (isdanni) [inline docs/syntax.rst:192]:
```suggestion
Function arguments are passed by value. So changes made inside function scope
won't affect the outside value in the caller:
```

File: docs/hello.rst (hunk 24 above)
Comment (archibate) [reply docs/hello.rst:137]:
```suggestion
  Everything decorated with ``@ti.kernel`` and ``@ti.func`` is in Taichi-scope
  and hence will be compiled by the Taichi compiler.

  Everything else is in Python-scope. They are simply Python native code.
```
Yes, I believe non-plural form is easier to read.

**Pull request merged**